Add AngularJS linky filter definition

aszmyd opened the pull request:
add `linky` filter definition for AngularJS: https://docs.angularjs.org/api/ngSanitize/filter/linky

thorn0 commented on angularjs/angular.d.ts:
```diff
@@ -746,6 +747,18 @@ declare namespace angular {
+         * @return Html-linkified and sanitized text.
+         * see https://docs.angularjs.org/api/ngSanitize/filter/linky
+         */
+        (text: string, target: string, attributes:Object|Function): string;
```
Please replace `Object` and `Function` with something more strongly typed.

thorn0 replied:
Also please mark the last two parameters as optional.

aszmyd replied:
Done. I've made only the `attributes` parameter optional as only it stays as optional in official docs: https://docs.angularjs.org/api/ngSanitize/filter/linky

thorn0 commented on angularjs/angular.d.ts:
```diff
@@ -756,7 +756,15 @@ declare namespace angular {
          * see https://docs.angularjs.org/api/ngSanitize/filter/linky
          */
-        (text: string, target: string, attributes:Object|Function): string;
+        <T>(text: string, target: string, attributes?:IFilterLinkyAttributesMapObject|IFilterLinkyAttributesFunc<T>): string;
```
Interfaces aren't needed for something as simple as `{[attribute: string]: string}`. Also the function shouldn't be generic and it returns an object not a string. 

> `function`: Takes the url as a parameter and returns a map of attributes

So the type of this parameter is:
```ts
attributes?: { [attribute: string]: string } | ((url: string) => { [attribute: string]: string })
```

aszmyd replied:
Right. Sorry for that. It should be as You've suggested now

thorn0 commented on angularjs/angular.d.ts:
```diff
@@ -746,6 +747,18 @@ declare namespace angular {
+         * @return Html-linkified and sanitized text.
+         * see https://docs.angularjs.org/api/ngSanitize/filter/linky
+         */
+        <T>(text: string, target: string, attributes?: { [attribute: string]: string } | ((url: string) => { [attribute: string]: string })): string;
```
Thanks. Please remove the `<T>` part, squash the commits, and we're good to go.

aszmyd replied:
Not sure what to do with all those rest filters definitions. `linky` is just additional definition that i've seen missing so i wanted to add it.

thorn0 commented:
Oh! I just realized these definitions belong to `angular-sanitize.d.ts`, not to `angular.d.ts`. Can you move them there? There are already typings for that filter, but they're not complete.

aszmyd commented:
Not sure what to do with all those rest filters definitions. `linky` is just additional definition that i've seen missing so i wanted to add it.

thorn0 commented:
`linky` is part of the `ngSanitize` module, not the core of Angular. That's why the type definitions for it should be in `angular-sanitize.d.ts`.

aszmyd commented:
Not sure if that was what You wanted. Im quite new to TS and not sure if its the right approach

thorn0 commented:
You're almost there. Don't modify angular.d.ts at all. Thanks to [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html), you can split the definition of the `angular.IFilterService` interface across multiple files. So, in `angular-sanitize.d.ts`, you can write something like this:

```ts
declare namespace angular {
	interface IFilterService {
		(name: 'linky'): angular.sanitize.filter.ILinky;
	}
}
```

Also it would be great if you added some test code using the added type definitions. See http://definitelytyped.org/guides/contributing.html#tests

thorn0 commented:
As for

> Im quite new to TS and not sure if its the right approach

Not a problem at all. Contributors are welcome. Also contributions are a great way to learn how the things work.

aszmyd commented:
@thorn0 Please check it now. I've moved the declaration to sanitize module, fix and write some more tests for the new signature.

thorn0 commented on angularjs/angular-sanitize-tests.ts:
```diff
@@ -1,10 +1,35 @@
+///////////////////////////////////////////////////////////////////////////////
+// Test sanitize service
+///////////////////////////////////////////////////////////////////////////////
+declare let $sanitizeService:ng.sanitize.ISanitizeService;
```
Please maintain the code style and use spaces after colons in type annotation.

thorn0 commented on angularjs/angular-sanitize.d.ts:
```diff
@@ -29,12 +29,25 @@ declare namespace angular.sanitize {
+         */
         interface ILinky {
-            (text: string, target?: string): string;
+            (text: string, target: "_blank" | "_self" | "_parent" | "_top", attributes?: { [attribute: string]: string } | ((url: string) => { [attribute: string]: string })): string;
```
> Window (`_blank|_self|_parent|_top`) _**or named frame**_ to open links in.

Which means it can be any string not just those listed.

thorn0 commented:
After you're done, please squash all the commits into one.

thorn0 commented on angularjs/angular-sanitize-tests.ts:
```diff
@@ -3,33 +3,30 @@
 ///////////////////////////////////////////////////////////////////////////////
-let shouldBeString:string;
-let testInputText:string = 'TEST';
+let shouldBeString : string;
```
The space is needed only after the colon, not before. Like you wrote here: 

```
(text: string, target: string...
     ^^
```

aszmyd commented:
Not Sure what You mean by squashing existing commits. I know GH has an tool to squash to one commit during merging but i'm not allowed to merge this. Also i'm not aware of any way to "clear" old commits and leave just one, squashed one. I've squashed them into one commit using this tutorial https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one/

aszmyd commented:
I can close this PR and open new one with only one commit inside if that helps

thorn0 commented:
See https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

So you have 14 commits in your branch. Do
```
git rebase -i HEAD~14
```
Pick the first one and squash the rest. Then edit the commit message and use `push --force` to overwrite the history on GitHub.

aszmyd commented:
I haven't made the `force` option last time. Now it looks like it's fine

thorn0 commented:
Looks good. :tada: Can be merged. :+1:

aszmyd commented:
Wonderful! Thanks for patience :)
